python/mxnet/autograd.py

```diff
@@ -411,16 +411,14 @@ class _Registry(object):
         """CustomOp registry."""
         def __init__(self):
             self.ref_holder = {}
-            self.counter = 0
-            self.lock = Lock()
+            self.counter = itertools.count()
```

Comment:
this object is already guarded by GIL so no other lock needed.

Reply:
There is nothing in the official documentation on itertools say that `itertools.count()` is thread-safe, although `itertools.count()` is thread-safe in CPython 3.8.

https://stackoverflow.com/questions/7083348/is-itertools-thread-safe

I think we can replace it with an integer, which is thread-safe.
https://stackoverflow.com/questions/6320107/are-python-ints-thread-safe

Reply:
thanks. for int in the context of unique counter, the race condition doesn't come from the mutation of the object itself, but the reference to it. so we should evaluate atomicity of the += operation instead. in this case, += on an integer variable is not thread safe.

given that the itertools.count() isn't guaranteed to be thread safe, I will revert it to use a lock instead.